Cabal/Distribution/PackageDescription/Check.hs

```diff
@@ -2136,6 +2116,41 @@ checkTarPath path
       ++ "Files with an empty name cannot be stored in a tar archive or in "
       ++ "standard file systems."
 
+-- | Check that setup dependencies, have proper upper bounds.
+-- @base@ and @Cabal@ upper dependencies are mandatory and emit
+-- error, all other packages emit warning.
+checkSetupVersions :: GenericPackageDescription -> [PackageCheck]
+checkSetupVersions pkg =
+    [ emitMessage $ unPackageName name
+    | (name, vr) <- Map.toList deps
+    , not (boundedAbove vr)
+    ]
+  where
+    criticalPkgs = ["Cabal", "base"]
+    deps = case typicalPkg pkg of
+      Right (pkgs', _) ->
+        Map.fromListWith intersectVersionRanges
+          [ (pname, vr)
+          | Just spi <- pure $ setupBuildInfo pkgs'
```

Comment:
I think you can use `maybeToList` here. Also, "spi" should be "sbi" ("bi" by convention means "build info").